File: backend/src/main/java/com/nanmuli/blog/domain/article/Article.java
package com.nanmuli.blog.domain.article;

import com.nanmuli.blog.shared.domain.BaseAggregateRoot;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;

import java.time.LocalDateTime;

@Getter
@Setter
@EqualsAndHashCode(callSuper = false)
public class Article extends BaseAggregateRoot<Long> {
    private static final long serialVersionUID = 1L;

    private String title;
    private String slug;
    private String content;
    private String contentHtml;
    private String summary;
    private String cover;
    private Long categoryId;
    private Long userId;
    private Integer viewCount;
    private Integer likeCount;
    private Integer wordCount;
    private Integer readingTime;
    private Integer status;
    private Boolean isTop;
    private Boolean isOriginal;
    private String originalUrl;
    private LocalDateTime publishTime;

    public ArticleId articleId() {
        return new ArticleId(this.id);
    }

    public ArticleStatus articleStatus() {
        return ArticleStatus.of(this.status);
    }

    public void publish() {
        this.status = ArticleStatus.PUBLISHED.getCode();
        this.publishTime = LocalDateTime.now();
    }

    public void draft() {
        this.status = ArticleStatus.DRAFT.getCode();
    }

    public void recycle() {
        this.status = ArticleStatus.RECYCLED.getCode();
    }

    public boolean isPublished() {
        return status != null && status == ArticleStatus.PUBLISHED.getCode();
    }

    public void calculateWordCount() {
        if (this.content != null) {
            this.wordCount = this.content.replaceAll("\\s+", "").length();
            this.readingTime = Math.max(1, this.wordCount / 300);
        }
    }
}

File: frontend/src/types/article.ts
import type { PageQuery } from './api'
import type { Category } from './category'

// 注意：ID 使用 string 类型，因为 JavaScript number 无法精确表示 64 位整数（Long）
// 后端使用雪花算法生成 ID，超过 2^53-1，前端必须用 string 保持精度
export interface Article {
  id: string
  title: string
  slug: string
  content: string
  contentHtml: string
  summary: string
  cover: string
  categoryId: string
  categoryName?: string  // 从category中提取
  category?: Category    // 完整分类信息
  categoryPath?: Category[]  // 分类层级路径（如：后端开发 > Java）
  viewCount: number
  likeCount: number
  wordCount: number
  readingTime: number
  status: number
  isTop: boolean
  isOriginal: boolean
  originalUrl: string
  publishTime: string
  createTime: string
  updateTime: string
}

export interface ArticleQuery extends PageQuery {
  categoryId?: string  // 按叶子分类筛选
  keyword?: string
  sort?: string
}

export interface ArticleListResult {
  list: Article[]
  total: number
}


export interface ArticleArchive {
  year: string
  month: string
  count: number
  articles?: ArticleSimple[]
}

export interface ArticleSimple {
  id: string
  title: string
  slug: string
  publishTime: string
}
